File: 04. Fall 2024/CS 4348 - Operating Systems Concepts/Project 01/Encryption.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Encryption {
    private static String passkey;

    public static void main(String[] args) {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(" ", 2);
                String command = parts[0];
                String argument = parts.length > 1 ? parts[1] : "";

                switch (command.toUpperCase()) {
                    case "PASSKEY":
                        passkey = argument;
                        System.out.println("RESULT");
                        break;
                    case "ENCRYPT":
                        if (passkey == null) {
                            System.out.println("ERROR Password not set");
                        } else {
                            System.out.println("RESULT " + encrypt(argument, passkey));
                        }
                        break;
                    case "DECRYPT":
                        if (passkey == null) {
                            System.out.println("ERROR Password not set");
                        } else {
                            System.out.println("RESULT " + decrypt(argument, passkey));
                        }
                        break;
                    case "QUIT":
                        return;
                    default:
                        System.out.println("ERROR Invalid command");
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static String encrypt(String text, String key) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            char c = (char) (((text.charAt(i) + key.charAt(i % key.length())) % 26) + 'A');
            result.append(c);
        }
        return result.toString();
    }

    private static String decrypt(String text, String key) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            char c = (char) (((text.charAt(i) - key.charAt(i % key.length()) + 26) % 26) + 'A');
            result.append(c);
        }
        return result.toString();
    }
}

File: 04. Fall 2024/CS 4348 - Operating Systems Concepts/Project 01/encryption.py
# encryption.py
import sys

passkey = None

def vigenere_encrypt(text, key):
    encrypted = []
    for i, char in enumerate(text):
        shift = ord(key[i % len(key)]) - ord('A')
        encrypted_char = chr((ord(char) - ord('A') + shift) % 26 + ord('A'))
        encrypted.append(encrypted_char)
    return ''.join(encrypted)

def vigenere_decrypt(text, key):
    decrypted = []
    for i, char in enumerate(text):
        shift = ord(key[i % len(key)]) - ord('A')
        decrypted_char = chr((ord(char) - ord('A') - shift + 26) % 26 + ord('A'))
        decrypted.append(decrypted_char)
    return ''.join(decrypted)

def main():
    global passkey
    while True:
        line = sys.stdin.readline().strip()
        if not line:
            continue
        command, *args = line.split(" ", 1)
        argument = args[0] if args else ""
        
        if command == "PASSKEY":
            passkey = argument.upper()
            print("RESULT")
        elif command == "ENCRYPT":
            if passkey is None:
                print("ERROR Password not set")
            else:
                encrypted_text = vigenere_encrypt(argument.upper(), passkey)
                print(f"RESULT {encrypted_text}")
        elif command == "DECRYPT":
            if passkey is None:
                print("ERROR Password not set")
            else:
                decrypted_text = vigenere_decrypt(argument.upper(), passkey)
                print(f"RESULT {decrypted_text}")
        elif command == "QUIT":
            break
        sys.stdout.flush()

if __name__ == "__main__":
    main()
